test/replaytests/choice.json

```diff
@@ -176,7 +176,34 @@
   },
   {
     "eventId": 12,
-    "timestamp": 1679427717321911295,
+    "timestamp": 1679427717321780254,
+    "eventType": "ActivityTaskStarted",
+    "version": 0,
+    "taskId": 5243011,
+    "activityTaskStartedEventAttributes": {
+      "scheduledEventId": 11,
+      "identity": "82203@agautam-NV709R969P@choiceGroup@41d230ae-253a-4d01-9079-322ef05c09fb",
+      "requestId": "ae2aad96-6588-4359-807b-a39a16f0896a",
+      "attempt": 0,
+      "lastFailureReason": ""
+    }
+  },
+  {
+    "eventId": 13,
+    "timestamp": 1679427717321780255,
+    "eventType": "ActivityTaskCompleted",
```

Comment:
why add the completed event?

though reading more of this file: I don't think this is a real history... "completed" implies "ended normally, i.e. by returning" but we don't create activity-task-scheduled events for activities executed during the final decision (since their result cannot be observed)...

Reply:
it wasn't a realistic history because it's missing the final decision task scheduled->started->completed events. I'm not quite getting why adding the activity completion looks invalid if that's what you meant

Reply:
well, there are only 2 ways workflows can end:
- return normally from the code (success or error)
- cut off externally due to timeout/termination/etc

and everything that affects ^ this occurs in transactions (no interleaving between these):
- a decision and all its operations go in as one chunk
- an external event *that the workflow can observe* goes in with the event and a decision-task-schedule to process it
- an external event that *cannot* be observed just gets appended
- cutoffs write one event and then end

---

in this history we originally had:
- decision task schedule->start->complete
- activity scheduled
- workflow complete
  - a normal one, which means it returned, which means this was the result of the most-recent decision task.

^ this is possible because, unlike child workflows, we don't "prune" activity-scheduled-s that are created in the final decision task.  (we should consider doing that tbh, or keeping both for easier troubleshooting)

So in this final decision task, we executed one activity, did not wait on it, and returned immediately.  Code like this would result in that:
```go
func work(ctx workflow.Context) error {
  // ...
  workflow.ExecuteActivity(ctx, "ignored") // non-blocking
  return nil
}
```

---

As of the current PR, we have this sequence:
- decision task schedule->start->complete
- activity scheduled
- activity started
- activity complete
- workflow complete (still normal)

which is not possible.

it would mean we blocked after scheduling (this is necessary to submit it to the server, which is a necessary prerequisite to the activity starting)... which is fine because `workflow.ExecuteActivity(...).Get(..)` would end the batch at "scheduled".

_however_.

after that, asynchronously, the activity was started (this can be an isolated event because it's not observable, that's fine), and then completed (this requires both complete and decision-task-scheduled because it _is_ observable, so already it's wrong), and then _also_ the workflow function returned on its own _without a decision task_ to learn about that result.

The only possible outcomes that include an activity that was started are things like either:
- activity started -> activity completed -> decision scheduled -> started -> completed -> workflow completed
- activity started -> workflow cut off externally (timeout, terminate) without a decision task

because "activity completed -> workflow completed" would be splitting the transaction that "activity completed" is involved in.

Reply:
tbh I'm kinda surprised it didn't panic with an illegal transition. But it would be an easy check to miss in the state machine too (particularly because it's more "a collection of many sub-state-machines" rather than enforcing rules in the spaces between them)